### PR TITLE
Added __iter__ and __contains__ method to pybamm.ParameterValues

### LINKED ISSUE #4438: Define `__iter__` and/or `__contains__` for `pybamm.ParameterValues`
### Description

The `pybamm.ParameterValues` is dict-like, and through its `__getitem__` and `__setitem__` methods allows dict-like key access. However, it does not support the `in` keyword, which for a dict will treat the keys as an iterable. This means that natural code like:

```python
if "Negative particle radius [m]" in parameter_values:
    # do something
```

will not work (generally raises a `KeyError`). It is currently necessary to instead write explicitly:

```python
if "Negative particle radius [m]" in parameter_values.keys():
    # do something
```

Defining the missing method `__iter__` (and `__contains__` if preferred) would allow `in` to work in logical test and for-loop contexts.

### Motivation

Code is cleaner if `pybamm.ParameterValues` looks like a souped-up dict always, not just sometimes. Key-value accessors are expected to iterate over their keys and to allow logical test on the presence/absence of keys. PyBaMM-containing code looks 'asymmetric' compared to code working with actual `dict` objects due to the requirement to call into `pybamm.ParameterValues.keys()` specifically.

Was there any historic reason for the decision to exclude these methods?

### Possible Implementation

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by AbhishekChaudharii**
# Description

Defined `__iter__ `and `__contains__ `method to `pybamm.ParameterValues` which would allow `in` to work in logical test and for-loop contexts. 

Fixes #4438 

## Type of change

- [x] New feature (non-breaking change which adds functionality)

# Key checklist:

- [x] No style issues: `$ pre-commit run` (or `$ nox -s pre-commit`) (see [CONTRIBUTING.md](https://github.com/pybamm-team/PyBaMM/blob/develop/CONTRIBUTING.md#installing-and-using-pre-commit) for how to set this up to run automatically when committing locally, in just two lines of code)
- [x] All tests pass: `$ python run-tests.py --all` (or `$ nox -s tests`)
- [ ] The documentation builds: `$ python run-tests.py --doctest` (or `$ nox -s doctests`)

You can run integration tests, unit tests, and doctests together at once, using `$ python run-tests.py --quick` (or `$ nox -s quick`).

## Further checks:

- [x] Code is commented, particularly in hard-to-understand areas
- [x] Tests added that prove fix is effective or that feature works

### CODE REVIEW

File: tests/unit/test_parameters/test_parameter_values.py [hunk 1]
```diff
@@ -1029,3 +1030,24 @@ def test_exchange_current_density_plating(self):
...
+            "Invalid key" not in parameter_values
+        ), "Non-existent key should not be found"
+
+    def test_iter_method(self):
+        """Test for __iter__ method to check if we can iterate over keys"""
+        parameter_values = ParameterValues(
+            values={"Negative particle radius [m]": 1e-6}
+        )
+        pv = [i for i in parameter_values]
+        print(pv)
```
Comment (agriyakhetarpal) [inline tests/unit/test_parameters/test_parameter_values.py:1052]:
```suggestion
```

It looks like this debugging statement was left in

**Comment by kratman** [comment]
Thanks @AbhishekChaudharii I just want to take a closer look at the tests, but it looks good to me so far

**Review by kratman — DISMISSED** [review]

**Comment by agriyakhetarpal** [comment]
@all-contributors please add @AbhishekChaudharii for tests

File: tests/unit/test_parameters/test_parameter_values.py [hunk 2]
```diff
@@ -1029,3 +1030,23 @@ def test_exchange_current_density_plating(self):
...
+            "Invalid key" not in parameter_values
+        ), "Non-existent key should not be found"
+
+    def test_iter_method(self):
+        """Test for __iter__ method to check if we can iterate over keys"""
+        parameter_values = ParameterValues(
+            values={"Negative particle radius [m]": 1e-6}
+        )
+        pv = [i for i in parameter_values]
+        assert len(pv) == 5, "Should have 5 keys "
```
Comment (arjxn-py) [inline tests/unit/test_parameters/test_parameter_values.py:1052]:
```suggestion
        assert len(pv) == 5, "Should have 5 keys"
```

Sorry for the nitpick :)

**Review by arjxn-py — DISMISSED** [review]
Thanks @AbhishekChaudharii, nice work 🚀

**Review by arjxn-py — APPROVED** [review]

**Review by agriyakhetarpal — APPROVED** [review]

**Comment by agriyakhetarpal** [comment]
Did this need a CHANGELOG entry?

**Comment by arjxn-py** [comment]
> Did this need a CHANGELOG entry?

Yes it should

**Comment by arjxn-py** [comment]
@kratman can you please add in #4466 maybe?

**Comment by agriyakhetarpal** [comment]
A new PR for this should be okay. @AbhishekChaudharii, please feel free to open one

**Pull request merged**